File: Assets/Script/Lib/Agent/Tool/EntryLease.lua
-- [ts]: EntryLease.ts
local ____exports = {} -- 1
local owner = "" -- 18
local runId -- 19
function ____exports.acquireEntryLease(id, entry) -- 20
	if owner ~= "" and owner ~= id then -- 20
		error("Dora entry runtime is busy with another Agent tool") -- 21
	end -- 21
	local status = entry.getCurrentEntryStatus() -- 22
	if status.running and (owner ~= id or status.runId ~= runId) then -- 22
		error("Dora entry runtime is in use; stop the current game before previewing") -- 23
	end -- 23
	owner = id -- 24
end -- 20
function ____exports.recordEntryLeaseRun(id, entry) -- 26
	if owner == id then -- 26
		runId = (entry.getCurrentEntryStatus().runId or 0) + 1 -- 27
	end -- 27
end -- 26
function ____exports.ownsEntryLease(id, entry) -- 29
	local status = entry.getCurrentEntryStatus() -- 30
	return owner == id and status.running and runId ~= nil and status.runId == runId -- 31
end -- 29
function ____exports.releaseEntryLease(id, entry) -- 33
	if owner ~= id then -- 33
		return nil -- 34
	end -- 34
	local cleanupError -- 35
	do -- 35
		local function ____catch(e) -- 35
			cleanupError = "failed to stop Agent preview: " .. tostring(e) -- 37
		end -- 37
		local ____try, ____hasReturned = pcall(function() -- 37
			if ____exports.ownsEntryLease(id, entry) and not entry.stop() then -- 37
				error("entry refused to stop") -- 36
			end -- 36
		end) -- 36
		if not ____try then -- 36
			____catch(____hasReturned) -- 36
		end -- 36
	end -- 36
	owner = "" -- 38
	runId = nil -- 38
	return cleanupError -- 39
end -- 33
return ____exports -- 33

File: Assets/Script/Lib/Agent/Tool/EntryLease.ts
// @preview-file off clear
export interface EntryStatus {
	success: boolean;
	running: boolean;
	runId?: number;
	kind?: string;
	fileName?: string;
	workDir?: string;
	projectRoot?: string;
	entryName?: string;
}
export interface DevEntryModule {
	allClear(this: void): void;
	stop(this: void): boolean;
	getCurrentEntryStatus(this: void): EntryStatus;
	enterEntryAsync(this: void, entry: {entryName: string; fileName: string; workDir: string; projectRoot: string; runKind: "agent_test"}): LuaMultiReturn<[boolean, string | undefined]>;
}
let owner = "";
let runId: number | undefined;
export function acquireEntryLease(id: string, entry: DevEntryModule): void {
	if (owner !== "" && owner !== id) error("Dora entry runtime is busy with another Agent tool");
	const status = entry.getCurrentEntryStatus();
	if (status.running && (owner !== id || status.runId !== runId)) error("Dora entry runtime is in use; stop the current game before previewing");
	owner = id;
}
// The lease predicts Entry's next run id as (current + 1). This relies on an
// invariant in Script.Dev.Entry: enterEntryAsync increments its run counter
// exactly once, synchronously at entry, and nothing else increments it.
export function recordEntryLeaseRun(id: string, entry: DevEntryModule): void {
	if (owner === id) runId = (entry.getCurrentEntryStatus().runId ?? 0) + 1;
}
export function ownsEntryLease(id: string, entry: DevEntryModule): boolean {
	const status = entry.getCurrentEntryStatus();
	return owner === id && status.running && runId !== undefined && status.runId === runId;
}
export function releaseEntryLease(id: string, entry: DevEntryModule): string | undefined {
	if (owner !== id) return undefined;
	let cleanupError: string | undefined;
	try { if (ownsEntryLease(id, entry) && !entry.stop()) error("entry refused to stop"); }
	catch (e) { cleanupError = `failed to stop Agent preview: ${tostring(e)}`; }
	owner = ""; runId = undefined;
	return cleanupError;
}
